stablehlo/transforms/StablehloConvertToSignless.cpp

```diff
@@ -0,0 +1,138 @@
+#include "stablehlo/transforms/Passes.h"
+
+namespace mlir {
+namespace stablehlo {
+
+#define GEN_PASS_DEF_STABLEHLOCONVERTTOSIGNLESSPASS
+#include "stablehlo/transforms/Passes.h.inc"
+
+namespace {
+
+// Generic pattern that rewrites any op by rewriting its operands and result
+// types. Regions are also rewritten.
+class ConvertToSignless : public ConversionPattern {
+ public:
+  ConvertToSignless(TypeConverter& typeConverter, MLIRContext* context)
+      : ConversionPattern(typeConverter, MatchAnyOpTypeTag{}, 0, context) {}
+
+  LogicalResult matchAndRewrite(
+      Operation* op, ArrayRef<Value> operands,
+      ConversionPatternRewriter& rewriter) const final {
+    SmallVector<Type> resultTypes;
+    if (failed(typeConverter->convertTypes(op->getResultTypes(), resultTypes)))
+      return failure();
+
+    auto* newOp = Operation::create(
+        op->getLoc(), op->getName(), resultTypes, operands, op->getAttrs(),
+        op->getPropertiesStorage(), op->getSuccessors(), op->getNumRegions());
+    for (auto regions : llvm::zip(op->getRegions(), newOp->getRegions())) {
+      Region& before = std::get<0>(regions);
+      Region& parent = std::get<1>(regions);
+      rewriter.inlineRegionBefore(before, parent, parent.end());
+      if (failed(rewriter.convertRegionTypes(&parent, *typeConverter)))
+        return failure();
+    }
+    rewriter.insert(newOp);
+    rewriter.replaceOp(op, newOp->getResults());
+    return success();
+  }
+};
+
+// A pattern that converts the type of the attribute used as an operand for
+// arith.constant
+class ConvertConstantToSignless
+    : public OpConversionPattern<arith::ConstantOp> {
+ public:
+  ConvertConstantToSignless(TypeConverter& typeConverter, MLIRContext* context)
+      : OpConversionPattern<arith::ConstantOp>(typeConverter, context) {}
+
+  LogicalResult matchAndRewrite(
+      arith::ConstantOp constantOp, arith::ConstantOpAdaptor adaptor,
+      ConversionPatternRewriter& rewriter) const override {
+    // We only care about unsigned integers
+    if (!adaptor.getValue().isa<DenseIntElementsAttr>()) return failure();
+
+    auto values = llvm::to_vector(
+        adaptor.getValue().cast<DenseIntElementsAttr>().getValues<APInt>());
+    Type type = typeConverter->convertType(constantOp.getType());
+    auto shapedType = type.dyn_cast<ShapedType>();
+    auto newValues = DenseIntElementsAttr::get(shapedType, values);
+
+    rewriter.replaceOpWithNewOp<arith::ConstantOp>(constantOp, newValues);
+    return success();
+  }
+};
+
+struct StablehloConvertToSignlessPass
+    : public impl::StablehloConvertToSignlessPassBase<
+          StablehloConvertToSignlessPass> {
+  using StablehloConvertToSignlessPassBase::StablehloConvertToSignlessPassBase;
+  void runOnOperation() override {
+    auto& context = getContext();
+    ConversionTarget target(context);
+
+    stablehlo::RemoveSignTypeConverter converter;
+    target.markUnknownOpDynamicallyLegal([&](auto op) {
```

Comment:
Looks like `unknown` and `arith::ConstantOp` are not covered by the existing test. Can we add a second function to that file which:

1. Uses tensors instead of memref (or is this memref specific, in which case we should modify the passes.td and consider moving next to the linalg lowering).
2. Uses a stablehlo op
3. Uses arith.constant

Reply:
Done